src/serac/physics/utilities/variational_form/weak_form.hpp

```diff
@@ -0,0 +1,500 @@
+   */
+  mfem::Vector& operator()(const mfem::Vector& input_T) const
+  {
+    Evaluation<Operation::Mult>(input_T, my_output_T_);
+    return my_output_T_;
+  }
+
+  /**
+   * @brief Obtains the gradients for all the constituent integrals
+   * @param[in] input_T The input vector
+   * @param[out] output_T The output vector
+   * @see Integral::GradientMult
+   */
+  virtual void GradientMult(const mfem::Vector& input_T, mfem::Vector& output_T) const
+  {
+    Evaluation<Operation::GradientMult>(input_T, output_T);
+  }
+
+  /**
+   * @brief Applies an essential boundary condition to the attributes specified by @a ess_attr
+   * @param[in] ess_attr The mesh attributes to apply the BC to
+   *
+   * @note This gets more interesting when having more than one trial space
+   */
+  void SetEssentialBC(const mfem::Array<int>& ess_attr)
+  {
+    static_assert(std::is_same_v<test, trial>, "can't specify essential bc on incompatible spaces");
+    test_space_->GetEssentialTrueDofs(ess_attr, ess_tdof_list_);
+  }
+
+private:
+  /**
+   * @brief Indicates whether to obtain values or gradients from a calculation
+   */
+  enum class Operation
+  {
+    Mult,
+    GradientMult
+  };
+
+  /**
+   * @brief Lightweight shim for mfem::Operator that produces the gradient of a @p WeakForm from a @p Mult
+   */
+  class Gradient : public mfem::Operator {
+  public:
+    /**
+     * @brief Constructs a Gradient wrapper that references a parent @p WeakForm
+     * @param[in] f The @p WeakForm to use for gradient calculations
+     */
+    Gradient(WeakForm& f) : mfem::Operator(f.Height()), form(f){};
+
+    virtual void Mult(const mfem::Vector& x, mfem::Vector& y) const override { form.GradientMult(x, y); }
+
+  private:
+    /**
+     * @brief The "parent" @p WeakForm to calculate gradients with
+     */
+    WeakForm<test(trial)>& form;
+  };
+
+  /**
+   * @brief Helper method for evaluation/gradient evaluation
+   * @tparam op Whether to obtain values or gradients
+   * @param[in] input_T The input vector
+   * @param[out] output_T The output vector
+   */
+  template <Operation op = Operation::Mult>
+  void Evaluation(const mfem::Vector& input_T, mfem::Vector& output_T) const
+  {
+    // get the values for each local processor
+    P_trial_->Mult(input_T, input_L_);
+
+    if (domain_integrals_.size() > 0) {
+      // get the values for each element on the local processor
+      G_trial_->Mult(input_L_, input_E_);
+
+      // compute residual contributions at the element level and sum them
+      output_E_ = 0.0;
+      for (auto& integral : domain_integrals_) {
+        if constexpr (op == Operation::Mult) {
+          integral.Mult(input_E_, output_E_);
+        }
+
+        if constexpr (op == Operation::GradientMult) {
+          integral.GradientMult(input_E_, output_E_);
+        }
+      }
+
+      // scatter-add to compute residuals on the local processor
+      G_test_->MultTranspose(output_E_, output_L_);
+    }
+
+#if 0
```

Comment:
Ahhh here's the comment. I would add a `TODO` so we know this should eventually get turned on.

Reply:
I tried to put everything related to the boundary integrals behind some conditional compilation, so it should be clearer why certain parts are disabled, until the mfem issues get addressed (or we find a workaround).